encrypt-config/FlowConfigEncryptor.cpp

```diff
@@ -61,114 +62,150 @@ std::vector<SensitiveProperty> listSensitiveProperties(const minifi::core::Proce
+    std::cout << "No sensitive property found with this component ID and property name.\n";
+    return {};
+  }
+  return {{sensitive_property_it->component_id, {{sensitive_property_it->property_name, property_value}}}};
+}
+
+std::unordered_map<minifi::utils::Identifier, std::unordered_map<std::string, std::string>> createOverridesForReEncryption(const std::vector<SensitiveProperty>& sensitive_properties) {
+  std::unordered_map<minifi::utils::Identifier, std::unordered_map<std::string, std::string>> overrides;
+  for (const auto& sensitive_property : sensitive_properties) {
+    overrides[sensitive_property.component_id].emplace(sensitive_property.property_name, sensitive_property.property_value);
+  }
+  return overrides;
+}
+
+}  // namespace
+
+namespace org::apache::nifi::minifi::encrypt_config::flow_config_encryptor {
+
+EncryptionRequest::EncryptionRequest(EncryptionType type) : type{type} {
+  gsl_Expects(type == EncryptionType::Interactive || type == EncryptionType::ReEncrypt);
+}
+
+EncryptionRequest::EncryptionRequest(std::string_view component_id, std::string_view property_name, std::string_view property_value)
+    : type{EncryptionType::SingleProperty},
+      component_id{component_id},
+      property_name{property_name},
+      property_value{property_value} {}
+
+void encryptSensitiveValuesInFlowConfig(const EncryptionKeys& keys, const std::filesystem::path& minifi_home, const std::filesystem::path& flow_config_path, const EncryptionRequest& request) {
+  const bool is_re_encrypting = keys.old_key.has_value();
+  if (is_re_encrypting && request.type != EncryptionType::ReEncrypt) {
+    throw std::runtime_error("Error: found .old key; please run --re-encrypt and then remove the .old key");
+  }
+  if (!is_re_encrypting && request.type == EncryptionType::ReEncrypt) {
+    throw std::runtime_error("Error: cannot re-encrypt without an .old key!");
+  }
+
+  const auto configure = std::make_shared<Configure>();
   configure->setHome(minifi_home);
   configure->loadConfigureFile(DEFAULT_NIFI_PROPERTIES_FILE);
 
-  bool encrypt_whole_flow_config_file = (configure->get(minifi::Configure::nifi_flow_configuration_encrypt) | minifi::utils::andThen(minifi::utils::string::toBool)).value_or(false);
-  auto encryptor = encrypt_whole_flow_config_file ? minifi::utils::crypto::EncryptionProvider::create(minifi_home) : std::nullopt;
-  auto filesystem = std::make_shared<minifi::utils::file::FileSystem>(encrypt_whole_flow_config_file, encryptor);
+  bool encrypt_whole_flow_config_file = (configure->get(Configure::nifi_flow_configuration_encrypt) | utils::andThen(utils::string::toBool)).value_or(false);
+  auto whole_file_encryptor = encrypt_whole_flow_config_file ? utils::crypto::EncryptionProvider::create(minifi_home) : std::nullopt;
+  auto filesystem = std::make_shared<utils::file::FileSystem>(encrypt_whole_flow_config_file, whole_file_encryptor);
+
+  auto sensitive_properties_encryptor = is_re_encrypting ?
```

Comment:
I think calling it decryptor would be more descriptive, since the old key is used for decrypting, before reencrypting with the new key.

Reply:
makes sense, changed in 762c4dfbaca0d4f7c9a40b3799017d8dfe48161b (although line 179 is going to look a bit weird now)